ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/CompletionDeduplicationInfoIT.scala

```diff
@@ -0,0 +1,182 @@
+import com.daml.ledger.api.testtool.infrastructure.Allocation._
+import com.daml.ledger.api.testtool.infrastructure.LedgerTestSuite
+import com.daml.ledger.api.testtool.infrastructure.participant.ParticipantTestContext
+import com.daml.ledger.api.testtool.suites.CompletionDeduplicationInfoIT._
+import com.daml.ledger.api.v1.command_service.SubmitAndWaitRequest
+import com.daml.ledger.api.v1.command_submission_service.SubmitRequest
+import com.daml.ledger.api.v1.commands.Command
+import com.daml.ledger.api.v1.completion.Completion
+import com.daml.ledger.api.v1.ledger_configuration_service.LedgerConfiguration
+import com.daml.ledger.client.binding.Primitive
+import com.daml.ledger.test.model.Test.Dummy
+import com.daml.lf.data.Ref
+import com.daml.platform.testing.WithTimeout
+import com.google.protobuf.duration.Duration
+import io.grpc.Status
+
+import scala.concurrent.duration.DurationInt
+import scala.concurrent.{ExecutionContext, Future}
+
+final class CompletionDeduplicationInfoIT(service: Service) extends LedgerTestSuite {
+
+  private val serviceName: String = service.productPrefix
+
+  override private[testtool] def name = serviceName + super.name
+
+  test(
+    shortIdentifier = s"CCDIIncludeDedupInfo-$serviceName",
+    description = s"Deduplication information is preserved in completions ($serviceName)",
+    allocate(SingleParty),
+  )(implicit ec => { case Participants(Participant(ledger, party)) =>
+    for {
+      config <- ledger.configuration()
+      optApplicationIdCompletion <- submitRequest(service, ledger, party, simpleCreate(party))
+      optSubmissionIdCompletion <- submitRequest(
+        service,
+        ledger,
+        party,
+        simpleCreate(party),
+        updateCommandServiceRequest = _.update(_.commands.submissionId := aSubmissionId),
+        updateCommandSubmissionServiceRequest = _.update(_.commands.submissionId := aSubmissionId),
+      )
+      optCompletionDeduplicationTime <- submitRequest(
+        service,
+        ledger,
+        party,
+        simpleCreate(party),
+        updateCommandServiceRequest = _.update(_.commands.deduplicationTime := aDeduplicationTime),
+        updateCommandSubmissionServiceRequest =
+          _.update(_.commands.deduplicationTime := aDeduplicationTime),
+      )
+    } yield {
+      assertApplicationIdIsPreserved(ledger.applicationId, optApplicationIdCompletion)
+      assertSubmissionIdIsGenerated(optApplicationIdCompletion)
+      assertDefaultDeduplicationTimeIsReported(config, optApplicationIdCompletion)
+      assertSubmissionIdIsPreserved(aSubmissionId, optSubmissionIdCompletion)
+      assertDeduplicationTimeIsPreserved(aDeduplicationTime, optCompletionDeduplicationTime)
+    }
+  })
+}
+
+private[testtool] object CompletionDeduplicationInfoIT {
+  sealed trait Service extends Serializable with Product
+  case object CommandService extends Service
+  case object CommandSubmissionService extends Service
+
+  private def submitRequest(
```

Comment:
I think it might be better to split this function into 2 and put each version into the proper service `case object`.

Reply:
Also, both cases look mostly the same, so it may be worth addressing this.

Reply:
3e404be but it's more verbose and structurally more complex. Let me know which one you prefer.

Reply:
Yeah, more verbose, as there's duplication that we won't be able to tackle easily. At least now the functions are smaller.  I'm fine with both.

Reply:
Unfortunately the two request types are only related at a `scalapb` level.